components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt

```diff
@@ -0,0 +1,199 @@
+            // Modernize certain language codes.
+            return when (val language = locale.language) {
+                "iw" -> "he"
+                "in" -> "id"
+                "ji" -> "yi"
+                else -> language
+            }
+        }
+    }
+
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Cached)
+
+    private lateinit var experiments: Experiments
+    private lateinit var dataDir: File
+    private var onExperimentUpdated: ((List<EnrolledExperiment>) -> Unit)? = null
+
+    init {
+        // Set the name of the native library
+        System.setProperty("uniffi.component.nimbus.libraryOverride", "nimbus")
+    }
+
+    /**
+     * Initialize the Nimbus SDK library.
+     *
+     * This should only be initialized once by the application.
+     *
+     * @param context [Context] to access application and device parameters.  As we cannot enforce
+     * through the compiler that the context pass to the initialize function is a Application
+     * Context, there could potentially be a memory leak if the initializing application doesn't
+     * comply.
+     *
+     * @return Returns a [Job] so that off-thread initialization can be monitored and interacted
+     * with.
+     */
+    fun initialize(
+        context: Context,
+        onExperimentUpdated: ((activeExperiments: List<EnrolledExperiment>) -> Unit)? = null
+    ) {
+        this.onExperimentUpdated = onExperimentUpdated
+
+        // Do initialization off of the main thread
+        scope.launch {
+            // Build Nimbus AppContext object to pass into initialize
+            val experimentContext = buildExperimentContext(context)
+
+            // Build a File object to represent the data directory for Nimbus data
+            dataDir = File(context.applicationInfo.dataDir, NIMBUS_DATA_DIR)
+
+            // Initialize Nimbus
+            experiments = Experiments(collectionName, experimentContext, dataDir.path, null)
+
+            // Get experiments
+            val activeExperiments = experiments.getActiveExperiments()
+
+            // Record enrollments in telemetry
+            recordExperimentTelemetry(activeExperiments)
+
+            // Invoke the callback with the list of active experiments for the consuming app to
+            // process.
+            onExperimentUpdated?.invoke(activeExperiments)
+        }
+    }
+
+    internal fun recordExperimentTelemetry(experiments: List<EnrolledExperiment>) {
+        // Call Glean.setExperimentActive() for each active experiment.
+        experiments.forEach {
+            // For now, we will just record the experiment id and the branch id. Once we can call
+            // Glean from the Nimbus-SDK Rust core, we will also record the targeting parameters
+            // and the bucketing parameters.
+            Glean.setExperimentActive(it.slug, it.branchSlug)
+        }
+
+        // Note, we cannot call setExperimentInactive unless we track the state of the
```

Comment:
A naive question, but this comment confused me in a couple of ways:

* I'd have thought that the rust component, which does persist stuff in the dbpath we pass as we initialize it, would mean that we could call `setExperimentInactive()` - ie, it's not clear to me if that is just a "todo" because `Experiments()` is yet to hook up db support, or a limitation that will remain for some time?

* The last 2 lines make it unclear to me whether all the sentences before it are relevant? ie, the tl;dr of this comment almost seems like it is "we can't do X, but we don't need to do X" - it doesn't seem documenting why we can't do something that we don't need to.

Reply:
Right now, Nimbus only checks for new experiments on application startup, similar to how Mako was doing it.  Since Glean doesn't persist the experiment annotation, it's up to the application to call the Glean experiments API with each launch, _and/or_ with any change to experiments after the first invocation.  Since we are only checking for experiments once (for now at least), there really isn't any point in calling `setExperimentInactive` until we are periodically updating experiments.  I admit, this API is a little weird for Glean, since it doesn't persist things it makes it just a little awkward to use.

Reply:
To clarify, this is more about dealing with a little weirdness in Glean rather than to do with the persistence of Nimbus.